File: proxy/media/expressions.py
"""A character's expression sprites: which labels exist, which filenames
carry them, and zipping the folder for export.

Export comes in two shapes (docs/FORKS_AND_EXTRAS_PLAN.md §2), and they are genuinely
different archives, not the same one with a different root:

* `zip_one` -- one character, flattened to basenames at the zip root. ST's
  importer (`getImageBuffers` behind the *Import Expressions Pack* button)
  keeps only `path.parse(entry.fileName).base`, so a flat zip is the shape
  that button expects; nesting would still load, since it flattens anyway.
* `zip_many` -- several characters, each under its own on-disk folder name
  (`<Name>_<gallery_id>`, exactly the string `resolve_folder` matches on). Not
  loadable through `upload-zip`, which targets one folder and would collapse
  every character into it -- this is for dropping straight back into
  `data/expressions/`.

Plain `zipfile` into an in-memory buffer; both shapes are asked for by hand,
not on a hot path, and a character's expressions run tens of megabytes, not
the hundreds that would make buffering the wrong call.
"""

from __future__ import annotations

import io
import zipfile
from pathlib import Path

# SillyTavern's GoEmotions sprite set, verbatim -- `DEFAULT_EXPRESSIONS` in its
# expressions extension, and the only labels an upload is allowed to carry
# (docs/FORKS_AND_EXTRAS_PLAN.md §2). The real 7,539-file corpus used exactly
# these 28 and no strays.
DEFAULT_EXPRESSIONS = frozenset(
    {
        "admiration",
        "amusement",
        "anger",
        "annoyance",
        "approval",
        "caring",
        "confusion",
        "curiosity",
        "desire",
        "disappointment",
        "disapproval",
        "disgust",
        "embarrassment",
        "excitement",
        "fear",
        "gratitude",
        "grief",
        "joy",
        "love",
        "nervousness",
        "neutral",
        "optimism",
        "pride",
        "realization",
        "relief",
        "remorse",
        "sadness",
        "surprise",
    }
)


def label_of(filename: str) -> str:
    """The expression a sprite filename names, derived exactly as ST's
    `src/endpoints/sprites.js` derives it: lowercase the stem (the name minus
    its last extension), then cut at the first `-` or `.` left in it. So
    `joy.webp`, `joy-1.webp` and `joy-_00004_.webp` are all `joy`.

    Never used to rename anything -- filenames are stored and exported
    verbatim. The frontend's `expressionLabel` is the same function; both
    exist because the label is both a display grouping and an upload gate.
    """
    stem = Path(filename).stem.lower()
    return stem.split("-")[0].split(".")[0]


def rejection_reason(filename: str) -> str | None:
    """Why this sprite is not storable in an expressions folder, or None. The
    label is what is checked, not the rest of the filename: ST ignores
    everything after the first separator, so `joy-whatever_042.webp` is a
    perfectly good `joy`."""
    label = label_of(filename)
    if label not in DEFAULT_EXPRESSIONS:
        return f"{label!r} is not one of SillyTavern's {len(DEFAULT_EXPRESSIONS)} expressions"
    return None


def _files(directory: Path) -> list[Path]:
    return sorted(
        p for p in directory.iterdir() if p.is_file() and not p.name.startswith(".")
    )


def zip_one(directory: Path) -> bytes:
    buffer = io.BytesIO()
    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as zf:
        for entry in _files(directory):
            zf.write(entry, arcname=entry.name)
    return buffer.getvalue()


def zip_many(folders: list[tuple[str, Path]]) -> bytes:
    """`folders` is `(folder name on disk, directory)` pairs -- the caller has
    already resolved each character's expressions folder."""
    buffer = io.BytesIO()
    with zipfile.ZipFile(buffer, "w", zipfile.ZIP_DEFLATED) as zf:
        for folder_name, directory in folders:
            for entry in _files(directory):
                zf.write(entry, arcname=f"{folder_name}/{entry.name}")
    return buffer.getvalue()

File: frontend/src/lib/expressions.ts
import type { GalleryFile } from './card'

/**
 * SillyTavern's GoEmotions sprite set, verbatim -- the labels a sprite
 * folder's filenames are grouped and validated against
 * (docs/FORKS_AND_EXTRAS_PLAN.md §2).
 */
export const DEFAULT_EXPRESSIONS = [
  'admiration',
  'amusement',
  'anger',
  'annoyance',
  'approval',
  'caring',
  'confusion',
  'curiosity',
  'desire',
  'disappointment',
  'disapproval',
  'disgust',
  'embarrassment',
  'excitement',
  'fear',
  'gratitude',
  'grief',
  'joy',
  'love',
  'nervousness',
  'neutral',
  'optimism',
  'pride',
  'realization',
  'relief',
  'remorse',
  'sadness',
  'surprise',
] as const

/**
 * The label a sprite filename carries, exactly as ST's own
 * `src/endpoints/sprites.js` derives it: lowercase the stem (the filename
 * minus its *last* extension only), then cut at the first `-` or `.` still in
 * it. `joy.webp`, `joy-1.webp` and `joy-_00004_.webp` are all `joy` --
 * verified against a real 7,539-file corpus at 100.00%. Never used to rename
 * anything; filenames are stored and exported verbatim, this is a display
 * concern only.
 */
export function expressionLabel(filename: string): string {
  const stem = filename.replace(/\.[^./]*$/, '').toLowerCase()
  return stem.split(/[-.]/)[0]
}

/** One display group: a label and the files that parsed to it. */
export interface ExpressionGroup {
  label: string
  files: GalleryFile[]
}

/**
 * Group a folder's files by parsed label, `neutral` first -- it is ST's
 * fallback sprite, so "does this character have a neutral?" is the question
 * worth answering at a glance. Everything else sorts alphabetically.
 */
export function groupExpressions(files: GalleryFile[]): ExpressionGroup[] {
  const groups = new Map<string, GalleryFile[]>()
  for (const file of files) {
    const label = expressionLabel(file.name)
    const list = groups.get(label)
    if (list) list.push(file)
    else groups.set(label, [file])
  }
  return [...groups.entries()]
    .map(([label, group]) => ({ label, files: group }))
    .sort((a, b) => {
      if (a.label === 'neutral') return -1
      if (b.label === 'neutral') return 1
      return a.label.localeCompare(b.label)
    })
}
